src/crowdin_client.cpp

```diff
@@ -135,150 +159,332 @@ CrowdinClient::~CrowdinClient() {}
+                const auto& dir = dirs[dirId];
+                path.push(dir.name);
+                dirId = dir.parentId;
+            }
+            string pathStr;
+            while(path.size()) {
+                pathStr += '/';
+                pathStr += path.top();
+                path.pop();
+            }
+            if(!pathStr.empty()) {
+                i.pathName = str::to_wstring(pathStr) + i.pathName;
+            }
+        }
+
+        //TODO: get more until all branches gotten (if more than 500)    
+        return m_api->get(url + "/branches?limit=500");
+    }).then([this, url, prj](json r) {
+        // Handle branches
+        map<int, string> branches;
+
+        for(const auto& i : r["data"]) {
+            const json& d = i["data"];
+            branches[d["id"]] = d["name"].get<string>();
+        }
+
+        for(auto& i : prj->files) {
+            if(i.branchId != NO_ID) {
+                i.pathName = "/" + str::to_wstring(branches[i.branchId]) + i.pathName;
+            }
+        }
+
+        return *prj;
+    });
 }
 
 
-dispatch::future<void> CrowdinClient::DownloadFile(const std::string& project_id,
-                                                   const std::wstring& file,
-                                                   const Language& lang,
+dispatch::future<void> CrowdinClient::DownloadFile(const long project_id,
+                                                   const long file_id,
+                                                   const wxString& file_name,
+                                                   const std::string& lang_tag,
                                                    const std::wstring& output_file)
 {
-    // NB: "export_translated_only" means the translation is not filled with the source string
-    //     if there's no translation, i.e. what Poedit wants.
-    auto url = "/api/project/" + project_id + "/export-file"
-                   "?json="
-                   "&export_translated_only=1"
-                   "&language=" + lang.LanguageTag() +
-                   "&file=" + http_client::url_encode(file);
-    return m_api->download(url, output_file);
+    cout << "\n\nGetting file URL: " << "/projects/" + std::to_string(project_id) + "/files/" + std::to_string(file_id) + "/download" << "\n\n";
+    return RefreshToken().then([=] () {
+    return m_api->post(
+        "projects/" + std::to_string(project_id) + "/translations/builds/files/" + std::to_string(file_id),
+        json_data(json({
+            { "targetLanguageId", lang_tag },
+            // for XLIFF files should be exported "as is" so set to `false`
+            { "exportAsXliff", !file_name.Lower().EndsWith(".xliff.xliff") }
+        })))
+        .then([this, output_file] (json r) {
+            string url = r["data"]["url"],
+                   proto(wxString(url).BeforeFirst(':').mb_str()),
+                   host(wxString(url).AfterFirst('/').AfterFirst('/').BeforeFirst('/').mb_str());
```

Comment:
If you must use wx APIs here, then there's `wxURI` that does proper parsing.

Are the download URLs arbitrary, disallowing sharing the client object?

Reply:
Done 199f93736a857f37a7a01438e9290a87558ca027